### ISSUE TITLE
Distribution of scheduling in random testing

### ISSUE THREAD

**Issue by aherrmann** [barrucadu/dejafu#257]
We encountered an issue with the distribution of scheduling in random testing.
A small example to illustrate the issue:

``` haskell
module Main where

import Control.Concurrent.Classy
import Control.Exception (Exception)
import Control.Monad (replicateM_, void, when)
import Data.Typeable (Typeable)
import System.Environment (getArgs)
import System.Random (mkStdGen)
import Test.DejaFu

newtype TestError = TestError String
  deriving (Show, Typeable)
instance Exception TestError

testfunc :: MonadConc m => Int -> m ()
testfunc repetitions = do
  mvShouldFail <- newMVarN "should-fail" False
  forkN "make-fail" $
    void $ swapMVar mvShouldFail True
  let failIfShould name = do
        shouldFail <- readMVar mvShouldFail
        when shouldFail $ throw (TestError name)
  replicateM_ repetitions $ failIfShould "A"
  replicateM_ repetitions $ failIfShould "B"


main :: IO ()
main = do
  [executions, repetitions] <- map read <$> getArgs
  print =<< dejafu "Systematic" notAlwaysSame (testfunc repetitions)
  print =<< dejafuWay (randomly (mkStdGen 0) executions) defaultMemType "Randomly" notAlwaysSame (testfunc repetitions)
```

The intention is to use the thread "make-fail" to simulate transient failures. It should inject a failure at any random point in the program. However, we find that the failure is not uniformly distributed over the program when we use random execution. In the above example systematic testing succeeds for essentially any number of repetitions (tried up to 100000000000). However, with random testing a very large number of executions can be necessary for a relatively small number of repetitions.

A failure looks like this:
```
[pass] Systematic
True
[fail] Randomly
    [TestError "A"] S0----P1---P0-P1---S0-
      1: make-fail

    [TestError "A"] S0----P1--P0-S1----S0--
      1: make-fail

    [TestError "A"] S0--------P1--P0-S1-P0-P1--P0-
      1: make-fail

    [TestError "A"] S0----P1---P0-P1---S0-
      1: make-fail

    [TestError "A"] S0------P1-P0--P1-P0-S1-P0--
      1: make-fail
    ...
False
```

The following table shows numbers of executions passed to `randomly` vs. the smallest number of repetitions required to make it fail. (Determined empirically):

| Executions | Min. repetitions to fail |
| --- | --- |
| 100 | 47 |
| 1000 | 99 |
| 10000 | 148 |
| 100000 | 185 |

The expectation would be that the failure is uniformly distributed and random testing also passes on the above example. Can this be achieved?

**Comment by barrucadu** [comment]
> The expectation would be that the failure is uniformly distributed and random testing also passes on the above example. Can this be achieved?

I don't think so.  The random choice is made at every scheduling point, so if the chance of scheduling `make-fail` is `p`, then the chance of skipping over all the `TestError "A"` repetitions is `(1 - p) ^ repetitions`, which is a very small number as `repetitions` grows.

I wonder if a smarter random algorithm would work better.  I think PCT would work well in this case, but that'd need me to have some way of estimating the length of the program.  I'll have to have a look through my paper collection again.

> In the above example systematic testing succeeds for essentially any number of repetitions (tried up to 100000000000).

I'm amazed.  Not that it passes, but that you ran a 200,000,000,000-step test case.  Did that take days?  Weeks?

---

So, leave this with me, and I'll think about how to make random scheduling better.  Sadly I don't have a solution right now.

**Comment by aherrmann** [comment]
> I'm amazed. Not that it passes, but that you ran a 200,000,000,000-step test case. Did that take days? Weeks?

Remarkably, it takes about two seconds. I didn't look into that before, but from 1000 `repetitions` upwards the execution time is actually more or less constant at about 2s on my machine (systematic and random testing at 100 `executions`). This is a bit suspicious, GHC must be doing something very clever with `replicateM_`. I haven't looked at the core, but I'm building with stack, which AFAIK defaults to `-O`. 

```
$ time stack exec dejafu-example 100 100000000000
...
real    0m1.954s
user    0m1.783s
sys     0m0.208s
```

I also tried manually unrolling `replicateM_` in the source. I.e.

``` haskell
  failIfShould "A"
  failIfShould "A"
  failIfShould "A"
  -- repeat ...
  failIfShould "B"
  failIfShould "B"
  failIfShould "B"
  -- repeat ...
```

At 100 such repetitions the execution takes 2.0s, at 1000 it takes 2.2s, at 3000 it takes 3.3s. For larger repetitions GHC takes too long to compile.

> I don't think so. The random choice is made at every scheduling point

Okay, that makes sense.

> So, leave this with me, and I'll think about how to make random scheduling better. Sadly I don't have a solution right now.

Thank you. It would be great to have a mechanism that allowed for these kinds of tests.

**Comment by barrucadu** [comment]
> This is a bit suspicious

Very suspicious.  And I've found the problem:

```haskell
print =<< resultsSet defaultWay defaultMemType (testfunc repetitions)
-- gives: fromList [Left Abort,Left (UncaughtException (TestError "A"))]
```

Since this has tripped a few people up now (including me until I looked at the results!) I clearly need to think about how failures are reported.  If I disable the length bound:

```haskell
print =<< resultsSet (systematically (defaultBounds { boundLength = Nothing })) defaultMemType (testfunc repetitions)
-- gives: internal error: Unable to commit 1048576 bytes of memory
```

Maybe the length bound should be disabled by default as well.  It's pretty arbitrary currently.

> I'm not familiar with PCT, and can't find it on google. What does PCT stand for?

"A Ran­dom­ized Sched­uler with Prob­ab­il­istic Guar­an­tees of Finding Bugs" ([pdf](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/paper-83.pdf))

It's a priority-based scheduler, where priorities are chosen randomly, and priority change points are sprinkled uniformly throughout the execution.  So it would handle the context switches happening uniformly throughout the execution, but it needs an estimate of how long the execution is.  That can probably be determined by maintaining a best guess, and doing a couple of trial runs first to get an initial estimate.
